File: src/utils/kitchenMode.ts
import {
  aisleIdFromProductId,
  KITCHEN_WARE_AISLES,
  KITCHEN_WARE_CATEGORY_ID,
  type KitchenAisleId,
} from "../data/kitchenWare";
import { productMatchesCategory } from "../data/categories";
import type { Product } from "../types/product";

export function isKitchenPath(pathname: string): boolean {
  if (pathname === "/kitchen" || pathname.startsWith("/kitchen/")) return true;
  if (pathname.includes("kitchen-ware")) return true;
  // Kitchen PDP keeps kitchen chrome (SKU ids are kitchen-{aisle}-…)
  return /^\/product\/kitchen-/.test(pathname);
}

export function isKitchenProduct(product: Product | null | undefined): boolean {
  if (!product) return false;
  return productMatchesCategory(product.category, KITCHEN_WARE_CATEGORY_ID);
}

export function kitchenAislePath(aisleId: string): string {
  return `/kitchen/aisle/${aisleId}`;
}

export function groupKitchenByAisle(products: Product[]) {
  const groups = new Map<string, Product[]>();
  for (const aisle of KITCHEN_WARE_AISLES) groups.set(aisle.id, []);

  for (const product of products) {
    const id = aisleIdFromProductId(product.id);
    if (!id) continue;
    const list = groups.get(id) ?? [];
    list.push(product);
    groups.set(id, list);
  }

  return KITCHEN_WARE_AISLES.map((aisle) => ({
    ...aisle,
    products: groups.get(aisle.id) ?? [],
  }));
}

export function filterKitchenProducts(products: Product[]): Product[] {
  return products.filter((product) =>
    productMatchesCategory(product.category, KITCHEN_WARE_CATEGORY_ID)
  );
}

export function filterKitchenAisle(
  products: Product[],
  aisleId: KitchenAisleId | string
): Product[] {
  return products.filter(
    (product) => aisleIdFromProductId(product.id) === aisleId
  );
}

export function isValidKitchenAisleId(id: string | undefined | null): id is KitchenAisleId {
  if (!id) return false;
  return KITCHEN_WARE_AISLES.some((a) => a.id === id);
}

File: mobile/lib/utils/kitchen_mode.dart
import '../data/kitchen_ware.dart';
import '../models/product.dart';
import 'categories.dart';

bool isKitchenPath(String location) {
  final path = Uri.tryParse(location)?.path ?? location;
  if (path == '/kitchen' || path.startsWith('/kitchen/')) return true;
  if (path.contains('kitchen-ware')) return true;
  // Kitchen PDP keeps kitchen chrome (SKU ids are kitchen-{aisle}-…).
  return RegExp(r'^/product/kitchen-').hasMatch(path);
}

bool isKitchenProduct(Product? product) {
  if (product == null) return false;
  if (product.id.startsWith('kitchen-')) return true;
  return productMatchesCategory(product.category, kitchenWareCategoryId);
}

List<Product> excludeKitchenProducts(List<Product> products) =>
    products.where((p) => !isKitchenProduct(p)).toList();

List<Product> filterKitchenProducts(List<Product> products) =>
    products.where(isKitchenProduct).toList();

List<Product> filterKitchenAisle(List<Product> products, String aisleId) =>
    products.where((p) => aisleIdFromProductId(p.id) == aisleId).toList();

/// Round-robin across aisles so home grids aren't one category (e.g. glasses).
List<Product> diversifyKitchenProducts(List<Product> products, {int limit = 24}) {
  if (products.isEmpty || limit <= 0) return const [];
  final groups = groupKitchenByAisle(products)
      .map((g) => g.products)
      .where((list) => list.isNotEmpty)
      .toList();
  if (groups.isEmpty) return products.take(limit).toList();

  final out = <Product>[];
  final seen = <String>{};
  var depth = 0;
  while (out.length < limit) {
    var added = false;
    for (final group in groups) {
      if (depth >= group.length) continue;
      final product = group[depth];
      if (seen.add(product.id)) {
        out.add(product);
        added = true;
        if (out.length >= limit) break;
      }
    }
    if (!added) break;
    depth++;
  }
  return out;
}

final _saucepanRe = RegExp(r'sauce\s*pans?', caseSensitive: false);

bool isSaucepanProduct(Product product) => _saucepanRe.hasMatch(product.title);

List<Product> filterSaucepans(List<Product> products) =>
    products.where(isSaucepanProduct).toList();

/// Saucepans first, then the rest (stable enough for shop grids).
List<Product> sortKitchenWithSaucepansFirst(List<Product> products) {
  final saucepans = <Product>[];
  final rest = <Product>[];
  for (final p in products) {
    if (isSaucepanProduct(p)) {
      saucepans.add(p);
    } else {
      rest.add(p);
    }
  }
  int byPopularity(Product a, Product b) =>
      (b.reviewCount * b.rating).compareTo(a.reviewCount * a.rating);
  saucepans.sort(byPopularity);
  rest.sort(byPopularity);
  return [...saucepans, ...rest];
}

class KitchenAisleGroup {
  const KitchenAisleGroup({
    required this.aisle,
    required this.products,
  });

  final KitchenAisle aisle;
  final List<Product> products;
}

List<KitchenAisleGroup> groupKitchenByAisle(List<Product> products) {
  final groups = <String, List<Product>>{
    for (final aisle in kitchenWareAisles) aisle.id: <Product>[],
  };

  for (final product in products) {
    final id = aisleIdFromProductId(product.id);
    if (id == null) continue;
    groups.putIfAbsent(id, () => <Product>[]).add(product);
  }

  return kitchenWareAisles
      .map(
        (aisle) => KitchenAisleGroup(
          aisle: aisle,
          products: groups[aisle.id] ?? const [],
        ),
      )
      .toList();
}
